src/c1_state_machine/p5_digital_cash.rs

```diff
@@ -48,9 +108,430 @@ impl StateMachine for DigitalCashSystem {
     type State = State;
     type Transition = CashTransaction;
 
-    fn next_state(starting_state: &State, t: &CashTransaction) -> State {
-        todo!("Exercise 1")
+    fn next_state(starting_state: &Self::State, t: &Self::Transition) -> Self::State {
+        match t {
+            CashTransaction::Mint { minter, amount} => handle_mint(starting_state.clone(), minter.clone(), *amount),
+            CashTransaction::Transfer { spends, receives } => handle_transfer(starting_state.clone(), spends.clone(), receives.clone())
+        }
+    }
+}
+
+fn handle_mint(starting_state: State, minter: User, amount: u64) -> State {
+    let mut next_state = starting_state;
+    let bill = Bill { owner: minter.clone(), amount: amount, serial: next_state.next_serial() };
+    next_state.insert_bill(bill);
+    next_state.increment_serial();
+    next_state
+}
+
+fn handle_transfer(starting_state: State, spends: Vec<Bill>, receives: Vec<Bill>) -> State {
+    if spends.is_empty() || receives.is_empty() {
+        return starting_state;
+    }
+    let mut next_state = starting_state.clone();
+
+    {
+        let input_set: BTreeSet<_> = spends.iter().collect();
+        if input_set.len() < spends.len() {
+            return starting_state;
+        }
+    }
+
+    {
+        let output_set: BTreeSet<_> = receives.iter().collect();
+        if output_set.len() < receives.len() {
+            return starting_state;
+        }
+    }
+
+    let mut total_input: u64 = 0;
+    let mut total_output: u64 = 0;
+
+    for input in spends.iter() {
+        let amount = input.amount();
+        let bill = input.clone();
+        if !next_state.contains_bill(bill) {
+            return starting_state;
+        }
+
+        if total_input.checked_add(amount).is_none() {
+            return starting_state;
+        }
```

Comment:
This is checking for an overflow right? I wonder if we should add a test to make sure this is handled correctly. Or if we want to keep this simple and not worry about it, you could just do normal addition here.

Reply:
Ya I can add one. Thanks for the catch